src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs

```diff
@@ -14,152 +14,152 @@
-        /// <returns>The MsalUiRequiredException if there is one, null, otherwise.</returns>
-        internal /* for testing */ static MsalUiRequiredException? FindMsalUiRequiredExceptionIfAny(Exception exception)
-        {
-            MsalUiRequiredException? msalUiRequiredException = exception as MsalUiRequiredException;
-            if (msalUiRequiredException != null)
-            {
-                return msalUiRequiredException;
-            }
-            else if (exception.InnerException != null)
-            {
-                return FindMsalUiRequiredExceptionIfAny(exception.InnerException);
-            }
-            else
-            {
-                return null;
-            }
-        }
-
-        private static bool IsAjaxRequest(HttpRequest request)
-        {
-            return string.Equals(request.Query[Constants.XRequestedWith], Constants.XmlHttpRequest, StringComparison.Ordinal) ||
-                string.Equals(request.Headers[Constants.XRequestedWith], Constants.XmlHttpRequest, StringComparison.Ordinal);
-        }
-    }
+	/// <summary>
+	/// Filter used on a controller action to trigger incremental consent.
+	/// </summary>
+	/// <example>
+	/// The following controller action will trigger.
+	/// <code>
+	/// [AuthorizeForScopes(Scopes = new[] {"Mail.Send"})]
+	/// public async Task&lt;IActionResult&gt; SendEmail()
+	/// {
+	/// }
+	/// </code>
+	/// </example>
+	public class AuthorizeForScopesAttribute : ExceptionFilterAttribute
+	{
+		/// <summary>
+		/// Scopes to request.
+		/// </summary>
+		public string[]? Scopes { get; set; }
+
+		/// <summary>
+		/// Key section on the configuration file that holds the scope value.
+		/// </summary>
+		public string? ScopeKeySection { get; set; }
+
+		/// <summary>
+		/// Azure AD B2C user flow.
+		/// </summary>
+		public string? UserFlow { get; set; }
+
+		/// <summary>
+		/// Allows specifying an AuthenticationScheme if OpenIdConnect is not the default challenge scheme.
+		/// </summary>
+		public string? AuthenticationScheme { get; set; }
+
+		/// <summary>
+		/// Handles the <see cref="MsalUiRequiredException"/>.
+		/// </summary>
+		/// <param name="context">Context provided by ASP.NET Core.</param>
+		public override void OnException(ExceptionContext context)
+		{
+			if (context != null)
+			{
+				MsalUiRequiredException? msalUiRequiredException = FindMsalUiRequiredExceptionIfAny(context.Exception);
+				if (msalUiRequiredException != null &&
+					IncrementalConsentAndConditionalAccessHelper.CanBeSolvedByReSignInOfUser(msalUiRequiredException))
+				{
+					// the users cannot provide both scopes and ScopeKeySection at the same time
+					if (!string.IsNullOrWhiteSpace(ScopeKeySection) && Scopes != null && Scopes.Length > 0)
+					{
+						throw new InvalidOperationException(
+							string.Format(
+								CultureInfo.InvariantCulture,
+								IDWebErrorMessage.ProvideEitherScopeKeySectionOrScopes,
```

Comment:
The error messages should really be localized in order to promote the most inclusive api experiences.

Reply:
Once localized, InvariantCulture is probably not the correct culture to use.